### ISSUE TITLE
Suggestion: Variable recall cost and expensive early djinns

### ISSUE THREAD

**Issue by Slayer95** [inferno8/wesnoth-To_Lands_Unknown#26]
Under the Burning Suns recently implemented a new feature to modify units recall cost depending on the unit cost and campaign difficulty (https://github.com/wesnoth/wesnoth/pull/7450). That implementation has its merits and might be worthy of testing because of the following issues.

1. Mystical Djinn and Summons Masters are way too viable in a monopolistic sense. There is basically no reason to advance Neutral or Heavy Summoners. Early djinn? Locked. Early rhamis? You get a HS for free in S04. Drain? Tons of mechanical, magical, and undead/destroyer foes. Rushing to get a Summon Master even rewards the player with a free AMLA.
2. Recalling Carpet Riders, Lv1 Elementals feels like a huge waste of gold. A large cost for dimensional gates (Mehir's "collection") doesn't make sense flavor-wise either.

On the other hand, I am not <i>a priori</i> exhilarated by the huge costs such a proposal implies for the highest difficulty (equivalent to recruiting a brand new unit).

Other than that, I believe that implementing variable recall cost would make early djinns a possibility. Given that the cost of <i>summoning</i> decreases the further the player progresses in the campaign, it wouldn't be a stretch to make the costs of <i>recalling</i> djinn (summoned by a Neutral Summoner) large in the first few scenarios. It could even be said that even Mehir's army wants to rip him off when recalling djinns. After Mehir becomes the leader of Tar-Tabar, he could confiscate (just to use badass terms) those djinn, so that their overcost is gone.

**Comment by inferno8** [comment]
Introducing Jinn earlier in the story would be a huge change, which would alter balance of many scenarios. On the other hand your points are very valid + it could make room for more "jinn advice" which many players find very cool about these units. 

I believe it could work along with the recall cost trick you mentioned, but it needs to be carefully implemented and thoroughly tested. Lore-wise all early acquired jinn would be owned not by Mehir but by the army of Sud-Affar and rented to him for a (recall) fee. Once Mehir becomes the leader of Tar-Tabar, he would automatically get ownership of those jinn ...and that's it. It could work.

In case of point 2. that also makes sense as long the change in recall cost isn't too big.

**Comment by Slayer95** [comment]
As a proof of concept, I'm gonna run S22 North Pole in Hard, with the following changes.
- ``recall_cost = 4*sqrt(recruit_cost)``
- _After_ that adjustment, decrease recruit costs of all four types of Lv1 elementals to their summoning costs (since Mehir's summon won't work with mobile keep; also, Mehir effectively gets the fast summon AMLA, but much better).
- Base gold increased from 400g to 600g. My expectation is that, starting from S10, the average recall cost would be ~25g, and it would later peak at ~30g, so the 50% increase in base gold compensates it. In detail, new recall costs would be:

![imagen](https://user-images.githubusercontent.com/4925744/227758960-89ae2339-101f-4123-8ca8-776d4c392417.png)

While I would like to take out my entire army of Summons Masters and Mystical Jinnis, they'd require 40g and 39g, respectively, which I might not have. So, let's see :]

(*) Cost of Dimensional Gates II will still need to be lowered, wbu making it equal to DG1 = 14g?

**Comment by Slayer95** [comment]
Went with full high-level units (plus a single small Earth Elemental) and it resulted in a true carnage. However, it shouldn't affect my chances in the final scenario.

A slight increase in gold or an increase in income/turn should result in a highly beneficial stream of low-level cannon fodder.

![imagen](https://user-images.githubusercontent.com/4925744/227779855-2fee88a9-785d-41e4-83a8-f17f4f36b330.png)

**Comment by Slayer95** [comment]
In order to get some momentum and feedback, maybe this could be implemented as a fourth difficulty level, what do you think?

**Comment by inferno8** [comment]
The new recall costs generated with your formula look good to me.

> Cost of Dimensional Gates II will still need to be lowered, wbu making it equal to DG1 = 14g?

Sounds reasonable.

> In order to get some momentum and feedback, maybe this could be implemented as a fourth difficulty level, what do you think?

Currently I have no plans of adding a fourth difficulty level, but if that helps with balancing and fine tuning the new recall cost mechanics then I might reconsider it.

**Comment by inferno8** [comment]
I have removed jinn confiscation mechanics and adjusted jinn recall cost in 37a7c12942d12d05bf424198f2eaa9cc140c894e 
New jinn advice has been implemented in 7820db454f86563bd0e3b1bb84395580ee2eb7bc

Currently the jinn overcost equals `+8g`. This means the player needs to spend 40g to recall a Wonderful Jinni. A hefty but fair price for a very mobile unit capable of regeneration, drain, very good cth and ability to foresee the future. Btw, his advancement, Mystical Jinni, is considered one of the best and most powerful units in all UMC.

The jinni overcost needs to be thoroughly tested. If this solution (recall cost modification) fails to achieve desired results, the second option is to implement a higher upkeep cost for all jinn.
